### DISCUSSION TITLE
Constrained dynamics with Kane method

### DISCUSSION THREAD

**Discussion by Simone2021space** [sympy/sympy#24417] [Q&A]
Hi,
when I model a multibody system (e.g. 2 links connected by joints), how is constrained dynamics handled by Kane method in sympy?
With a penalty based method or using Lagrange multipliers?

**Comment by Peter230655** [comment]
You have some system with a holonomic constraint, which cannot be solved explicitly for the generalized coordinate(s), correct?

If correct, what I do is to convert this holonomic constraint into velocity constraints, then there is a formalism in sympy.physics.mechanics to handle this.
What I mean is this:
You have a constraint
hol = .... ( and it must be: hol = 0)
then I form:
holx = me.dot(hol, N.x)			N = frame of inertia
holy = me.dot(hol, N.y)
holz = me.dot(hol, N.z)

Then

holxdt = holx.diff(t)
holydt = holy.diff(t)
holzdt = holz.diff(t)

These are linear equations for the generalized speeds, which can always be solved explicitly for the dependent speeds, and used when you set up KanesMethod.

This works fine, as long as the system is not too complicated.
If this does not work, at least I am stuck.

**Reply by moorepants** [reply]
>  If this does not work, at least I am stuck.

You can setup Lagrange multipliers as per the standard method if the substitution methods do not work due to symbolic complexity.

**Comment by moorepants** [comment]
There are two chapters in my book that explains how to deal with holonomic and nonholonomic constraints using Kane's method: https://moorepants.github.io/learn-multibody-dynamics/

**Comment by Peter230655** [comment]
Dear Jason,

I know your lecture to some extend - and I enjoyed it! :-)
I specifically meant equation 175.
Using sympy.physics.mechanics, I can set up
*f_k* and  *f_d* two different ways:

- pretending the system has no holonomic constraint
- assuming, as it does, it has a holonomic constraint, which in convert
into speed    constraints as per your lecture.

*f_h* is clear.

I assume, that if I want to use *dae*, I should find MM and force using the
speed constraints, but I am not 100% sure.

Merry Christmas,

Peter

On Fri 23. Dec 2022 at 17:48 Jason K. Moore ***@***.***>
wrote:

> There are two chapters in my book that explains how to deal with holonomic
> and nonholonomic constraints using Kane's method:
> https://moorepants.github.io/learn-multibody-dynamics/
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sympy/sympy/discussions/24417#discussioncomment-4487465>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AT5MQUVRRVKR3QTEHWR2IFDWOXJVHANCNFSM6AAAAAATHARFDQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Best regards,

Peter Stahlecker
